Code/client/Services/Generic/CombatService.cpp

```diff
@@ -0,0 +1,277 @@
+    request.DeferInitialization = acEvent.DeferInitialization;
+    request.ForceConeOfFire = acEvent.ForceConeOfFire;
+
+#if TP_SKYRIM64
+    request.UnkBool1 = acEvent.UnkBool1;
+    request.UnkBool2 = acEvent.UnkBool2;
+#else
+    request.ConeOfFireRadiusMult = acEvent.ConeOfFireRadiusMult;
+    request.Tracer = acEvent.Tracer;
+    request.IntentionalMiss = acEvent.IntentionalMiss;
+    request.Allow3D = acEvent.Allow3D;
+    request.Penetrates = acEvent.Penetrates;
+    request.IgnoreNearCollisions = acEvent.IgnoreNearCollisions;
+#endif
+
+    m_transport.Send(request);
+}
+
+void CombatService::OnNotifyProjectileLaunch(const NotifyProjectileLaunch& acMessage) const noexcept
+{
+    ModSystem& modSystem = World::Get().GetModSystem();
+
+    auto remoteView = m_world.view<RemoteComponent, FormIdComponent>();
+    const auto remoteIt = std::find_if(std::begin(remoteView), std::end(remoteView), [remoteView, Id = acMessage.ShooterID](auto entity)
+    {
+        return remoteView.get<RemoteComponent>(entity).Id == Id;
+    });
+
+    if (remoteIt == std::end(remoteView))
+    {
+        spdlog::warn("Shooter with remote id {:X} not found.", acMessage.ShooterID);
+        return;
+    }
+
+    FormIdComponent formIdComponent = remoteView.get<FormIdComponent>(*remoteIt);
+
+#if TP_SKYRIM64
+    Projectile::LaunchData launchData{};
+#else
+    ProjectileLaunchData launchData{};
+#endif
+
+    launchData.pShooter = Cast<TESObjectREFR>(TESForm::GetById(formIdComponent.Id));
+
+    launchData.Origin.x = acMessage.OriginX;
+    launchData.Origin.y = acMessage.OriginY;
+    launchData.Origin.z = acMessage.OriginZ;
+
+    const uint32_t cProjectileBaseId = modSystem.GetGameId(acMessage.ProjectileBaseID);
+    launchData.pProjectileBase = TESForm::GetById(cProjectileBaseId);
+
+#if TP_SKYRIM64
+    const uint32_t cFromWeaponId = modSystem.GetGameId(acMessage.WeaponID);
+    launchData.pFromWeapon = Cast<TESObjectWEAP>(TESForm::GetById(cFromWeaponId));
+#endif
+
+#if TP_FALLOUT4
+    Actor* pShooter = Cast<Actor>(launchData.pShooter);
+    pShooter->GetCurrentWeapon(&launchData.FromWeapon, 0);
+#endif
+
+    const uint32_t cFromAmmoId = modSystem.GetGameId(acMessage.AmmoID);
+    launchData.pFromAmmo = Cast<TESAmmo>(TESForm::GetById(cFromAmmoId));
+
+    launchData.fZAngle = acMessage.ZAngle;
+    launchData.fXAngle = acMessage.XAngle;
+    launchData.fYAngle = acMessage.YAngle;
+
+    const uint32_t cParentCellId = modSystem.GetGameId(acMessage.ParentCellID);
+    launchData.pParentCell = Cast<TESObjectCELL>(TESForm::GetById(cParentCellId));
+
+    const uint32_t cSpellId = modSystem.GetGameId(acMessage.SpellID);
+    launchData.pSpell = Cast<MagicItem>(TESForm::GetById(cSpellId));
+
+    launchData.eCastingSource = (MagicSystem::CastingSource)acMessage.CastingSource;
+
+    launchData.iArea = acMessage.Area;
+    launchData.fPower = acMessage.Power;
+    launchData.fScale = acMessage.Scale;
+
+    launchData.bAlwaysHit = acMessage.AlwaysHit;
```

Comment:
Could be better to name them the same?

Reply:
I considered that, but the member names in LaunchData are closer to the actual names used in the in-game struct. Could do either way.